File: wanted/src/main/java/wanted/domain/product/service/ProductService.java
package wanted.domain.product.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import wanted.common.exception.CustomException;
import wanted.common.exception.code.GlobalExceptionCode;
import wanted.domain.brand.entity.Brand;
import wanted.domain.brand.repository.BrandRepository;
import wanted.domain.category.entity.Category;
import wanted.domain.category.repository.CategoryRepository;
import wanted.domain.product.dto.Pagination;
import wanted.domain.product.dto.ProductCategoryRequest;
import wanted.domain.product.dto.ProductCreateRequest;
import wanted.domain.product.dto.ProductCreateResponse;
import wanted.domain.product.dto.ProductImageRequest;
import wanted.domain.product.dto.ProductListResponse;
import wanted.domain.product.dto.ProductOptionGroupRequest;
import wanted.domain.product.dto.ProductOptionRequest;
import wanted.domain.product.dto.ProductSearchCondition;
import wanted.domain.product.dto.SimpleProductResponse;
import wanted.domain.product.entity.Product;
import wanted.domain.product.entity.ProductCategory;
import wanted.domain.product.entity.ProductDetail;
import wanted.domain.product.entity.ProductImage;
import wanted.domain.product.entity.ProductOption;
import wanted.domain.product.entity.ProductOptionGroup;
import wanted.domain.product.entity.ProductPrice;
import wanted.domain.product.entity.ProductTag;
import wanted.domain.product.repository.ProductCategoryRepository;
import wanted.domain.product.repository.ProductDetailRepository;
import wanted.domain.product.repository.ProductImageRepository;
import wanted.domain.product.repository.ProductOptionGroupRepository;
import wanted.domain.product.repository.ProductOptionRepository;
import wanted.domain.product.repository.ProductPriceRepository;
import wanted.domain.product.repository.ProductQueryRepository;
import wanted.domain.product.repository.ProductRepository;
import wanted.domain.product.repository.ProductTagRepository;
import wanted.domain.seller.entity.Seller;
import wanted.domain.seller.repository.SellerRepository;
import wanted.domain.tag.entity.Tag;
import wanted.domain.tag.repository.TagRepository;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ProductService {
    private final ProductRepository productRepository;
    private final SellerRepository sellerRepository;
    private final BrandRepository brandRepository;
    private final ProductDetailRepository productDetailRepository;
    private final ProductPriceRepository productPriceRepository;
    private final ProductCategoryRepository productCategoryRepository;
    private final CategoryRepository categoryRepository;
    private final ProductOptionGroupRepository productOptionGroupRepository;
    private final ProductOptionRepository productOptionRepository;
    private final ProductImageRepository productImageRepository;
    private final TagRepository tagRepository;
    private final ProductTagRepository productTagRepository;
    private final ProductQueryRepository productQueryRepository;


    /*
1. Product                (먼저 저장 → 기본 ID 생성됨)
2. ProductDetail          (1:1)
3. ProductPrice           (1:1)
4. ProductCategory        (1:N)
5. ProductOptionGroup     (1:N)
6. ProductOption          (N:1)
7. ProductImage           (1:N, option FK nullable)
8. ProductTag             (1:N)
 */
    @Transactional
    public ProductCreateResponse createProduct(ProductCreateRequest productCreateRequest){
        Seller seller = sellerRepository.findById(productCreateRequest.sellerId())
                .orElseThrow(() -> new CustomException(GlobalExceptionCode.RESOURCE_NOT_FOUND, resourceNotFoundDetails("Seller", productCreateRequest.sellerId())));

        Brand brand = brandRepository.findById(productCreateRequest.brandId())
                .orElseThrow(() -> new CustomException(GlobalExceptionCode.RESOURCE_NOT_FOUND, resourceNotFoundDetails("Brand", productCreateRequest.brandId())));

        Product product = Product.from(productCreateRequest, seller, brand);
        productRepository.save(product);

        ProductDetail productDetail = ProductDetail.from(productCreateRequest.detail(), product);
        productDetailRepository.save(productDetail);

        ProductPrice productPrice = ProductPrice.from(productCreateRequest.price(), product);
        productPriceRepository.save(productPrice);

        saveProductCategory(productCreateRequest, product);
        saveProductOption(productCreateRequest, product);
        saveProductImage(productCreateRequest, product);
        saveProductTag(productCreateRequest, product);

        return ProductCreateResponse.of(product);
    }

    private void saveProductCategory(ProductCreateRequest productCreateRequest, Product product) {
        for (ProductCategoryRequest productCategoryRequest : productCreateRequest.categories()) {
            Category category = categoryRepository.findById(productCategoryRequest.categoryId())
                    .orElseThrow(() -> new CustomException(GlobalExceptionCode.RESOURCE_NOT_FOUND, resourceNotFoundDetails("Category", productCategoryRequest.categoryId())));
            ProductCategory productCategory = ProductCategory.from(productCategoryRequest, product, category);
            productCategoryRepository.save(productCategory);
        }
    }

    private void saveProductOption(ProductCreateRequest productCreateRequest, Product product) {
        for (ProductOptionGroupRequest productOptionGroupRequest : productCreateRequest.optionGroups()) {
            ProductOptionGroup productOptionGroup = ProductOptionGroup.from(productOptionGroupRequest, product);
            productOptionGroupRepository.save(productOptionGroup);

            for (ProductOptionRequest productOptionRequest : productOptionGroupRequest.options()) {
                ProductOption productOption = ProductOption.from(productOptionRequest, productOptionGroup);
                productOptionRepository.save(productOption);
            }
        }
    }

    private void saveProductImage(ProductCreateRequest productCreateRequest, Product product) {
        if (productCreateRequest.images() == null) return;
        for (ProductImageRequest productImageRequest : productCreateRequest.images()) {
            ProductOption productOption = null;

            if (productImageRequest.optionId() != null) {
                productOption = productOptionRepository.findById(productImageRequest.optionId())
                        .orElseThrow(() -> new CustomException(GlobalExceptionCode.RESOURCE_NOT_FOUND, resourceNotFoundDetails("Option", productImageRequest.optionId())));
            }
            ProductImage productImage = ProductImage.from(productImageRequest, product, productOption);
            productImageRepository.save(productImage);
        }
    }

    private void saveProductTag(ProductCreateRequest productCreateRequest, Product product) {
        if (productCreateRequest.tags() == null) return;
        for (Long tagId : productCreateRequest.tags()) {
            Tag tag = tagRepository.findById(tagId)
                    .orElseThrow(() -> new CustomException(GlobalExceptionCode.RESOURCE_NOT_FOUND, resourceNotFoundDetails("Tag", tagId)));

            ProductTag productTag = ProductTag.from(product, tag);
            productTagRepository.save(productTag);
        }
    }

    @Transactional(readOnly = true)
    public ProductListResponse getFilteredProducts(ProductSearchCondition condition) {
        int page = Optional.ofNullable(condition.page()).orElse(1) - 1;
        int perPage = Optional.ofNullable(condition.perPage()).orElse(10);
        Pageable pageable = PageRequest.of(page, perPage, parseSort(condition.sort()));

        Page<Product> result = productQueryRepository.search(condition, pageable);
        List<SimpleProductResponse> items = result.getContent().stream()
                .map(SimpleProductResponse::from)
                .toList();

        return new ProductListResponse(
                items,
                new Pagination(result.getTotalElements(), result.getTotalPages(), result.getNumber() + 1, result.getSize())
        );
    }

    private Sort parseSort(String sortString) {
        if (sortString == null || sortString.isBlank()) {
            return Sort.by(Sort.Direction.DESC, "createdAt");
        }

        return Sort.by(
                Arrays.stream(sortString.split(","))
                        .map(s -> {
                            String[] parts = s.split(":");
                            String field = parts[0];
                            Sort.Direction direction = (parts.length > 1 && parts[1].equalsIgnoreCase("asc"))
                                    ? Sort.Direction.ASC : Sort.Direction.DESC;
                            return new Sort.Order(direction, field);
                        }).toList()
        );
    }


    private Map<String, Object> resourceNotFoundDetails(String type, Object id) {
        return Map.of("resourceType", type, "resourceId", id);
    }
}

File: src/product/application/services/Product.service.ts
import { Inject, Injectable, NotFoundException } from "@nestjs/common";
import { EntityManager } from "typeorm";

import { IBaseRepository } from "@libs/domain/repositories";
import { Product, Product_Detail, Product_Image, Product_Price } from "@product/domain/entities";
import {
  FilterDTO,
  ProductCatalogDTO,
  ProductCategoryDTO,
  ProductInputDTO,
  ProductOptionGroupDTO,
  ProductSummaryDTO,
  ProductTagDTO,
} from "../dto";

@Injectable()
export default class ProductService {
  constructor(
    private readonly entity_manager: EntityManager,
    @Inject("IProductRepository")
    private readonly repository: IBaseRepository<Product | ProductSummaryDTO | ProductCatalogDTO>,
    @Inject("IProductDetailRepository")
    private readonly product_detail_repository: IBaseRepository<Product_Detail>,
    @Inject("IProductPriceRepository")
    private readonly product_price_repository: IBaseRepository<Product_Price>,
    @Inject("IProductCategoryRepository")
    private readonly product_category_repository: IBaseRepository<ProductCategoryDTO>,
    @Inject("IProductOptionGroupRepository")
    private readonly product_option_group_repository: IBaseRepository<ProductOptionGroupDTO>,
    @Inject("IProductImageRepository")
    private readonly product_image_repository: IBaseRepository<Product_Image>,
    @Inject("IProductTagRepository")
    private readonly product_tag_repository: IBaseRepository<ProductTagDTO>,
  ) {}

  async register({
    detail,
    price,
    categories,
    option_groups,
    images,
    tags: tag_ids,
    seller_id,
    brand_id,
    ...product
  }: ProductInputDTO) {
    // 상품 등록 트랜잭션 처리
    const product_entity = await this.entity_manager.transaction(async (manager) => {
      // 상품 등록
      const product_entity = await this.repository.with_transaction(manager).save({
        ...product,
        seller_id,
        brand_id,
      });
      const { id: product_id } = product_entity;

      // 상품 상세 등록
      await this.product_detail_repository.with_transaction(manager).save({
        ...detail,
        product_id,
      });

      // 상품 가격 등록
      await this.product_price_repository.with_transaction(manager).save({
        ...price,
        product_id,
      });

      // 상품 카테고리 등록
      await this.product_category_repository
        .with_transaction(manager)
        .saves(categories.map((category) => ({ ...category, product_id })));

      // 상품 옵션 등록
      await this.product_option_group_repository
        .with_transaction(manager)
        .saves(option_groups.map((group) => ({ ...group, product_id })));

      // 상품 이미지 등록
      await this.product_image_repository
        .with_transaction(manager)
        .saves(images.map((image) => ({ ...image, product_id })));

      // 상품 태그 등록
      await this.product_tag_repository
        .with_transaction(manager)
        .saves(tag_ids.map((tag_id) => ({ tag_id, product_id })));

      return product_entity;
    });
    // 상품 등록 결과 반환
    return (({ id, name, slug, created_at, updated_at }) => ({
      id: id!,
      name,
      slug,
      created_at: created_at!,
      updated_at: updated_at!,
    }))(product_entity);
  }

  async find_all({ page = 1, per_page = 10, sort, ...rest }: FilterDTO) {
    const [sort_field, sort_order] = sort?.split(":") ?? ["created_at", "DESC"];

    const items = (await this.repository.find_by_filters({
      page,
      per_page,
      sort_field,
      sort_order,
      ...rest,
    })) as ProductSummaryDTO[];

    // 페이지네이션 요약 정보
    const pagination = {
      total_items: items.length,
      total_pages: Math.ceil(items.length / (per_page ?? 10)),
      current_page: page ?? 1,
      per_page: per_page ?? 10,
    };

    return { items, pagination };
  }

  async find(id: number) {
    const product = await this.repository.find_by_id(id);

    if (!product) {
      throw new NotFoundException({
        message: "요청한 리소스를 찾을 수 없습니다.",
        details: { resourceType: "Product", resourceId: id },
      });
    }
    return product as ProductCatalogDTO;
  }

  async edit(
    product_id: number,
    { detail, seller_id, brand_id, price, categories, ...product }: ProductInputDTO,
  ) {
    const is_updated = await this.entity_manager.transaction(async (manager) => {
      // 상품 디테일 업데이트
      await this.product_detail_repository
        .with_transaction(manager)
        .update({ ...detail, product_id });

      // 상품 가격 업데이트
      await this.product_price_repository
        .with_transaction(manager)
        .update({ ...price, product_id });

      // 상품 카테고리 업데이트
      for (const category of categories) {
        await this.product_category_repository
          .with_transaction(manager)
          .update({ ...category, product_id });
      }

      // 상품 제품 업데이트
      return this.repository.with_transaction(manager).update(
        {
          seller_id,
          brand_id,
          ...product,
        },
        product_id,
      );
    });

    if (!is_updated) {
      throw new NotFoundException({
        message: "요청한 리소스를 찾을 수 없습니다.",
        details: { resourceType: "Product", resourceId: product_id },
      });
    }

    const updated_product = await this.repository.find_by_id(product_id);

    return (({ id, name, slug, updated_at }) => ({
      id: id!,
      name,
      slug,
      updated_at: updated_at!,
    }))(updated_product!);
  }

  async remove(id: number) {
    const is_deleted = await this.repository.delete(id);

    if (!is_deleted) {
      throw new NotFoundException({
        message: "요청한 리소스를 찾을 수 없습니다.",
        details: { resourceType: "Product", resourceId: id },
      });
    }
  }
}
